### PR TITLE
fix: Query params trigger the trailingSlash error in preview mode

### LINKED ISSUE #8942: Query params trigger the `trailingSlash` error in preview mode
### Astro Info

```block
Astro                    v3.4.0
Node                     v20.9.0
System                   Linux (x64)
Package Manager          pnpm
Output                   static
Adapter                  none
Integrations             @astrojs/mdx
                         @astrojs/solid-js
                         pagefind
```

### If this issue only occurs in one browser, which browser is a problem?

n/a

### Describe the Bug

When navigating to for example `http://localhost:4321/search/?q=hask` while running `astro preview` an error is shown:

image

### What's the expected result?

The page should be shown normally

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-rj7etb?file=src%2Flayouts%2FLayout.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by rishi-raj-jain**
## Changes

So this fixes #8942. It basically tweaks the pathname logic to only look at pathname instead of (currently) pathname with query params.

## Testing

I think with the current testing suite this was already covered. I didn't particularly test this change.

## Docs

I don't think we need a documentation change for this as it was an internal configuration that broke the expectation from preview server.

### CODE REVIEW

File: packages/astro/src/core/preview/vite-plugin-astro-preview.ts [hunk 1]
```diff
@@ -24,7 +24,8 @@ export function vitePluginAstroPreview(settings: AstroSettings): Plugin {
 					return;
 				}
 
-				const pathname = stripBase(req.url!, base);
+				let strippedPathname = stripBase(req.url!, base);
```
Comment (florian-lefebvre) [inline packages/astro/src/core/preview/vite-plugin-astro-preview.ts:27]:
```suggestion
				const strippedPathname = stripBase(req.url!, base);
```
Can this be a const? Doesn't look like it's reassigned

File: packages/astro/src/core/preview/vite-plugin-astro-preview.ts [hunk 2]
```diff
@@ -24,7 +24,8 @@ export function vitePluginAstroPreview(settings: AstroSettings): Plugin {
 					return;
 				}
 
-				const pathname = stripBase(req.url!, base);
+				const strippedPathname = stripBase(req.url!, base);
+        const pathname = new URL(strippedPathname, 'https://a.b').pathname
```
Comment (florian-lefebvre) [inline packages/astro/src/core/preview/vite-plugin-astro-preview.ts:28]: I'm not familiar with the codebase but what is this `a.b` doing?

Comment (rishi-raj-jain) [reply packages/astro/src/core/preview/vite-plugin-astro-preview.ts:28]: It's anything, we don't care about the domain. We just extract the pathname but need the URL method to correctly parse the pathname.

File: .changeset/tidy-peas-accept.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'astro': patch
+---
+
+fix preview server pathname parse
```
Comment (bluwy) [inline .changeset/tidy-peas-accept.md:5]:
```suggestion
Fixes preview server `trailingSlash` handling for request URLs with query strings
```

**Review by bluwy — APPROVED** [review]

**Pull request merged**